src/v/cluster/types.cc

```diff
@@ -323,6 +328,7 @@ adl<cluster::topic_configuration>::from(iobuf_parser& in) {
     cfg.properties.retention_bytes = adl<tristate<size_t>>{}.from(in);
     cfg.properties.retention_duration
       = adl<tristate<std::chrono::milliseconds>>{}.from(in);
+    cfg.properties.recovery = adl<std::optional<bool>>{}.from(in);
```

Comment:
can you comment on the safety here wrt to forward/backward compatibility and our upgrade/restart procedure? do we need to handle this by inspecting the version of the structure?

Reply:
I think that the issue here is that encoding a std::optional (even if it is not set) will modify the serialized format. so in this case new data is added to the resulting serialized byte representation of the topic_configuration.

new code may fail to deserialize old versions of the structure (probably an issue for us) and old code may fail to deserialize new versions of the data (perhaps not an issue, but a concern when upgrading).

to complicate things slightly, the serialization of the topic_configuration appears to be embedded in the serialization of other types. i'm not immediately seeing a version being written for these types so it might be tricky.

does that sound right @mmaslankaprv ?